c10/util/BFloat16.h

```diff
@@ -0,0 +1,57 @@
+#pragma once
+
+// Defines the bloat16 type (brain floating-point). This representation uses
+// 1 bit for the sign, 8 bits for the exponent and 7 bits for the mantissa.
+
+#include <c10/macros/Macros.h>
+#include <cmath>
+#include <cstring>
+
+namespace c10 {
+
+namespace detail {
+  inline bool isSmallEndian() {
+    int num = 1;
+    return *(char *)&num == 1;
+  }
+
+  inline float f32_from_bits(uint16_t src) {
+    float res = 0;
+    uint32_t tmp = src;
+    if (isSmallEndian()) {
```

Comment:
This is wrong on big-endian systems. Please remove all the little-endian checks.

The TensorFlow code needs the little-endian check because they are reinterpret_cast a float* to uint16* which depends on the byte-order. I have no idea whey they implemented it that way -- it seems unnecessarily complicated as well as violating strict-aliasing rules.

However, none of the operations here depend on byte order, except the memcpy assumes the ordering matches between uint32 and float.

https://github.com/tensorflow/tensorflow/blob/9590c4c32dd4346ea5c35673336f5912c6072bf2/tensorflow/core/framework/bfloat16.cc#L36

Reply:
See https://gcc.godbolt.org/z/ARxFXh

Note that with powerpc64 clang (trunk) [big endian] z and y do not match.

Reply:
you are right. thanks for the godbolt link.
retested: https://gcc.godbolt.org/z/yEF0sM